lambda/lambda.ml

```diff
@@ -673,79 +673,101 @@ let rec make_sequence fn = function
    Assumes that the image of the substitution is out of reach
    of the bound variables of the lambda-term (no capture). *)
 
-let subst update_env s lam =
-  let rec subst s lam =
-    let remove_list l s =
-      List.fold_left (fun s (id, _kind) -> Ident.Map.remove id s) s l
-    in
+let subst update_env ?(on_bound_variable = fun id -> id) s lam =
+  let bv id l =
+    let id' = on_bound_variable id in
+    id', Ident.Map.add id id' l in
+  let bvs ids l =
+    List.fold_right (fun (id, kind) (ids', l) ->
+        let id', l = bv id l in
+        ((id', kind) :: ids' , l)
+      ) ids ([], l) in
+  let rec subst s l lam =
     match lam with
-    | Lvar id as l ->
-        begin try Ident.Map.find id s with Not_found -> l end
+    | Lvar id as lam ->
+        begin match Ident.Map.find_opt id l with
+          | Some id' -> Lvar id'
+          | None ->
+             begin match Ident.Map.find_opt id s with
+               | Some lam' -> lam'
+               | None -> lam
+             end
+        end
     | Lconst _ as l -> l
     | Lapply ap ->
-        Lapply{ap with ap_func = subst s ap.ap_func;
-                      ap_args = subst_list s ap.ap_args}
+        Lapply{ap with ap_func = subst s l ap.ap_func;
+                      ap_args = subst_list s l ap.ap_args}
     | Lfunction lf ->
-        let s =
-          List.fold_right
-            (fun (id, _) s -> Ident.Map.remove id s)
```

Comment:
Why remove that ?

Reply:
The `remove` are removing the bound variables from the substitution, so that only free variables remain in the domain of the substitution. My idea was that having two environments of free and bound variables obviates this: I can always tell when something is a bound variable (would, before, have been removed from the map) by looking in `l`.

In the `Tvar` case I lookup in the bound-variable environment first, and only then in the free-variable environment. This means that there is no lookup in `s` for bound variables, so the `remove` would make no difference.

Reply:
Actually, supporting the odd situations you mention was an explicit addition of #1606, supposedly for the purpose of "a (later) patch about let-rec compilation". I have cold feet about restricting the specification of the function again; I don't need it for my implementation, and I don't want to be responsible for potential breakages that could occur.

Reply:
(I can see how you could end up in those odd situations if you inline a recursive definition within itself without freshening its bound identifiers. Interesting, the *present* PR provides an easy way to freshen those bound identifiers. In any case, I think that any such considerations could be left for another PR, possibly by someone else.)